### ISSUE TITLE
"Tab does not have the required permissions for this operation"

### ISSUE THREAD

**Issue by itslittlejohn** [OfficeDev/microsoft-teams-library-js#69]
I am trying to figure out how to successfully call `microsoftTeams.authentication.getAuthToken`. I keep getting back `"Tab does not have the required permissions for this operation"`. 

Stepping into that method (not this library, but the teams window source), I see that the permission level for my personal scope tab is `1` (prompt) instead of `2` (full). In my manifest, I have `permissions: ['identity', 'messageTeamMembers']` and there don't seem to be any other permissions settings that I can set. 

I have the developer preview setting enabled and am the admin for the test tenant I have set up. I realize that `getAuthToken` and `getUser` aren't documented on the official site yet. Is that intentional since they aren't yet working or maybe I'm just missing something?

Also, I cannot find any documentation listing the valid resource URIs for the `getAuthToken` call. I get the same error response though for both `getAuthToken` and `getUser`.

Thanks.

**Comment by ydogandjiev** [comment]
Hi @itslittlejohn, the use of both getAuthToken and getUser are currently limited to internal 1P Microsoft apps which is why they are not documented on MSDN. We are looking into bringing SSO support for 3P AAD-based apps (possibly via these same APIs sometime in the future). In the meantime, please follow the instructions here to implement popup-based authentication for your tab using a standard OAuth flow:
https://msdn.microsoft.com/en-us/microsoft-teams/auth

**Comment by itslittlejohn** [comment]
OK, thanks for clearing that up. I kept seeing oblique references to "silent authentication" here and there and assumed that this was the right way to do it. 

For example: https://blogs.msdn.microsoft.com/richard_dizeregas_blog/2017/02/08/microsoft-teams-and-oauth-in-custom-tab/

Is this actually no longer supported? If it is, could you point me towards the correct documentation for that?

I was able to get a standard OAuth flow working, but was more interested in the supposed "silent" approach I was reading about.

**Comment by ydogandjiev** [comment]
@itslittlejohn, the closest you can get to an SSO experience in your tab today is by:
1. Silently redirecting to login.microsoftonline.com with the ```prompt=none``` and ```login_hint=<user_upn>``` parameters. You can read more about these parameters [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-v2-protocols-implicit#get-access-tokens).
2. In most cases where the user already has an active AAD session in the current context, AAD will re-direct back to you with the security code.
3. If AAD redirects back to you with an error instead then you should call ```microsoftTeams.authentication.authenticate``` and trigger the same OAuth flow inside of the popup (this time **without** the ```prompt=none``` parameter).

The above-described solution will minimize the number of times a user is prompted for their credentials but not completely eliminate them - the first time a user signs into your app they will have to consent (regardless of whether there is an active AAD session or not); also, the first time a user uses any AAD-based tab in the Teams desktop client they will have to re-enter their credentials.

To completely eliminate these unnecessary prompts, we are working on a solution similar to the way [SSO works for Office Add-ins](https://dev.office.com/docs/add-ins/develop/sso-in-office-add-ins) but that's not quite ready yet.

**Comment by itslittlejohn** [comment]
Thanks @ydogandjiev. I implemented this approach, though I haven't yet encountered a scenario where the `prompt=none` method works. Maybe it is because of the new Microsoft login pages? I always get a `login_required` error.

**Comment by billbliss** [comment]
Check out this article (though you've already uncovered most of it): https://techcommunity.microsoft.com/t5/Microsoft-Teams-Blog/Authentication-SSO-and-Microsoft-Graph-in-Microsoft-Teams-Tabs/ba-p/125366. 

In particular, see this: https://github.com/OfficeDev/microsoft-teams-sample-complete-node/blob/master/src/views/tab-auth/silent-start.hbs. 

That example does not use `prompt=none` but it is mostly silent (you see a popup flash briefly but then it disappears). The only other difference between that and what you are doing is `scope=openid+profile`. 

This is the best that's possible now until we implement [Office 365 add-in style](https://docs.microsoft.com/en-us/office/dev/add-ins/develop/sso-in-office-add-ins) SSO as @ydogandjiev mentioned.

**Comment by ydogandjiev** [comment]
@itslittlejohn can you please provide the AAD URL that you are redirecting to? Are you adding the login_hint parameter?

**Comment by itslittlejohn** [comment]
Thanks for the help, guys, and sorry for the slow response. We realized that the issue was using `common` instead of the tenant upn in the redirect URL. I also started to explore using MSAL instead of doing all of this manually, but we ran into some [issues](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/222) with using it from within teams, so once we get back around to getting auth fully implemented, we'll likely use the approach outlined above.
